File: app/src/main/java/com/metahumanz/pacilread/util/FileAssetHelper.java
package com.metahumanz.pacilread.util;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;
import java.util.UUID;

public final class FileAssetHelper {
    private FileAssetHelper() {
    }

    public static File copyUriToFolder(Context context, Uri uri, String folderName, String prefix) throws IOException {
        Context appContext = context.getApplicationContext();
        File folder = new File(appContext.getFilesDir(), folderName);
        if (!folder.exists() && !folder.mkdirs()) {
            throw new IOException("无法创建目录: " + folder.getAbsolutePath());
        }

        String extension = extensionOf(resolveDisplayName(appContext, uri));
        File destination = new File(folder, safePrefix(prefix) + "_" + UUID.randomUUID() + extension);
        try (InputStream inputStream = appContext.getContentResolver().openInputStream(uri);
             FileOutputStream outputStream = new FileOutputStream(destination)) {
            if (inputStream == null) {
                throw new IOException("无法读取文件内容");
            }
            byte[] buffer = new byte[8192];
            int read;
            while ((read = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, read);
            }
        }
        return destination;
    }

    public static String resolveDisplayName(Context context, Uri uri) {
        ContentResolver resolver = context.getApplicationContext().getContentResolver();
        try (Cursor cursor = resolver.query(uri, null, null, null, null)) {
            if (cursor != null && cursor.moveToFirst()) {
                int index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                if (index >= 0) {
                    String value = cursor.getString(index);
                    if (value != null && !value.isBlank()) {
                        return value;
                    }
                }
            }
        } catch (Exception ignore) {
        }
        String fallback = uri.getLastPathSegment();
        return fallback == null || fallback.isBlank() ? "asset.bin" : fallback;
    }

    public static void deleteIfExists(String path) {
        if (path == null || path.isBlank()) {
            return;
        }
        File file = new File(path);
        if (file.exists()) {
            file.delete();
        }
    }

    private static String extensionOf(String fileName) {
        if (fileName == null) {
            return "";
        }
        int dotIndex = fileName.lastIndexOf('.');
        if (dotIndex < 0) {
            return "";
        }
        return fileName.substring(dotIndex).toLowerCase(Locale.ROOT);
    }

    private static String safePrefix(String prefix) {
        String value = prefix == null ? "asset" : prefix.trim().toLowerCase(Locale.ROOT);
        value = value.replaceAll("[^a-z0-9_-]+", "_");
        return value.isEmpty() ? "asset" : value;
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/util/FileAssetHelper.kt
package com.metahumanz.pacilread.util

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.Locale
import java.util.UUID

object FileAssetHelper {
    @JvmStatic
    @Throws(IOException::class)
    fun copyUriToFolder(context: Context, uri: Uri, folderName: String?, prefix: String?): File {
        val appContext = context.applicationContext
        val folder = File(appContext.filesDir, folderName!!)
        if (!folder.exists() && !folder.mkdirs()) throw IOException("无法创建目录: ${folder.absolutePath}")

        val extension = extensionOf(resolveDisplayName(appContext, uri))
        val destination = File(folder, "${safePrefix(prefix)}_${UUID.randomUUID()}$extension")
        val inputStream = appContext.contentResolver.openInputStream(uri)
        val outputStream = FileOutputStream(destination)
        inputStream.use { input ->
            outputStream.use { output ->
                if (input == null) throw IOException("无法读取文件内容")
                val buffer = ByteArray(8192)
                while (true) {
                    val read = input.read(buffer)
                    if (read == -1) break
                    output.write(buffer, 0, read)
                }
            }
        }
        return destination
    }

    @JvmStatic
    fun resolveDisplayName(context: Context, uri: Uri): String {
        val resolver = context.applicationContext.contentResolver
        try {
            resolver.query(uri, null, null, null, null).use { cursor ->
                if (cursor != null && cursor.moveToFirst()) {
                    val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (index >= 0) {
                        val value = cursor.getString(index)
                        if (!value.isNullOrBlank()) return value
                    }
                }
            }
        } catch (_: Exception) {
        }
        val fallback = uri.lastPathSegment
        return if (fallback.isNullOrBlank()) "asset.bin" else fallback
    }

    @JvmStatic
    fun deleteIfExists(path: String?) {
        if (path.isNullOrBlank()) return
        val file = File(path)
        if (file.exists()) file.delete()
    }

    private fun extensionOf(fileName: String?): String {
        if (fileName == null) return ""
        val dotIndex = fileName.lastIndexOf('.')
        if (dotIndex < 0) return ""
        return fileName.substring(dotIndex).lowercase(Locale.ROOT)
    }

    private fun safePrefix(prefix: String?): String {
        val value = (prefix?.trim()?.lowercase(Locale.ROOT) ?: "asset").replace(Regex("[^a-z0-9_-]+"), "_")
        return value.ifEmpty { "asset" }
    }
}
